vector/src/test/java/im/vector/app/features/onboarding/RegistrationActionHandlerTest.kt

```diff
@@ -0,0 +1,73 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package im.vector.app.features.onboarding
+
+import im.vector.app.test.fakes.FakeRegistrationWizard
+import im.vector.app.test.fakes.FakeSession
+import kotlinx.coroutines.test.runBlockingTest
+import org.amshove.kluent.shouldBeEqualTo
+import org.junit.Test
+import org.matrix.android.sdk.api.auth.registration.RegisterThreePid
+import org.matrix.android.sdk.api.auth.registration.RegistrationResult
+import org.matrix.android.sdk.api.auth.registration.RegistrationWizard
+
+private val A_SESSION = FakeSession()
+private val AN_EXPECTED_RESULT = RegistrationResult.Success(A_SESSION)
+private const val A_USERNAME = "a username"
+private const val A_PASSWORD = "a password"
+private const val AN_INITIAL_DEVICE_NAME = "a device name"
+private const val A_CAPTCHA_RESPONSE = "a captcha response"
+private const val A_PID_CODE = "a pid code"
+private const val EMAIL_VALIDATED_DELAY = 10000L
+private val A_PID_TO_REGISTER = RegisterThreePid.Email("an email")
+
+class RegistrationActionHandlerTest {
+
+    private val fakeRegistrationWizard = FakeRegistrationWizard()
+    private val registrationActionHandler = RegistrationActionHandler()
+
+    @Test
+    fun `when handling register action then delegates to wizard`() = runBlockingTest {
+        val cases = listOf(
+                case(RegisterAction.StartRegistration) { getRegistrationFlow() },
+                case(RegisterAction.CaptchaDone(A_CAPTCHA_RESPONSE)) { performReCaptcha(A_CAPTCHA_RESPONSE) },
+                case(RegisterAction.AcceptTerms) { acceptTerms() },
+                case(RegisterAction.RegisterDummy) { dummy() },
+                case(RegisterAction.AddThreePid(A_PID_TO_REGISTER)) { addThreePid(A_PID_TO_REGISTER) },
+                case(RegisterAction.SendAgainThreePid) { sendAgainThreePid() },
+                case(RegisterAction.ValidateThreePid(A_PID_CODE)) { handleValidateThreePid(A_PID_CODE) },
+                case(RegisterAction.CheckIfEmailHasBeenValidated(EMAIL_VALIDATED_DELAY)) { checkIfEmailHasBeenValidated(EMAIL_VALIDATED_DELAY) },
+                case(RegisterAction.CreateAccount(A_USERNAME, A_PASSWORD, AN_INITIAL_DEVICE_NAME)) {
+                    createAccount(A_USERNAME, A_PASSWORD, AN_INITIAL_DEVICE_NAME)
+                }
+        )
+
+        cases.forEach { testSuccessfulActionDelegation(it) }
+    }
+
+    private suspend fun testSuccessfulActionDelegation(case: Case) {
+        fakeRegistrationWizard.givenSuccessFor(result = A_SESSION, case.expect)
```

Comment:
@ouchadam I don't know if you have seen this comment? Not a blocker, but I just wanted to know your opinion about it.

Reply:
ah I somehow missed this :sweat_smile:  great point! yes, will definitely create a new instance per case

Reply:
for checking if the expected case is called I'm a little weary of using verifies on non side effect functions as they can cause false positives (although your example also includes checking the return value for extra security)

```kotlin
fun handle(wizard, action) {
  when(action) {
   foo -> 
         // wrong behaviour but a verify acceptTerms would still pass
         wizard.acceptTerms()
         wizard.dummy()
  }
}
```

I would lean towards the verify and result assertion being a little tooo similar as we need to return a value in order for the test to pass, if the wizard was triggering `Unit` side effects then I would 100% be for also verifying 

could be swayed but would prefer to include less code in the test if possible!

Reply:
I see you point. In fact, the reason I wanted to add this verification is to be sure we do not call any other methods on the wizard instance than the one that produces the result (to avoid calls which may produced side effects). I used `coVerifyAll` which means: "starts a verification block that should include all calls for coroutines".
But I will approve the PR, it is up to you if you want to add this verification.

Reply:
if the mocks are set with `relaxed=false (default)` then any non mocked functions will throw but someone could easily switch the setting to `relaxed=true`, we need to have some processes/rules~ :thinking: 

```kotlin
// with a non relaxed mock
val mock = mock<RegistrationWizard>(relaxed  = false)
coEvery { mock.acceptTerms() } returns RegistrationResult.Success(result)

mock.acceptTerms()
mock.dummy() // throws exception due to not being mocked
```

```kotlin
// with a relaxed mock
val mock = mock<RegistrationWizard>(relaxed  = true)
coEvery { mock.acceptTerms() } returns RegistrationResult.Success(result)

mock.acceptTerms()
mock.dummy() 

coVerifyAll { mock.acceptTerms() }  // fails due to dummy()
```

if I've fully understood your point (correct me if I'm wrong!), we're both aligned with trying to fail the tests when we call unexpected methods and it comes down to whether we want to explicitly call verify or rely on the lack of mocking to catch unexpected calls

Reply:
Yes we are aligned 😄 I have forgotten there was this `relaxed` parameter. I would say using `relaxed` makes this verification in an implicit way while with the `verify` pattern we explicitly says what we want to test. On my side, I would go for `verify` pattern (but we can keep the `relaxed = false` as well).

Reply:
I've included the verify https://github.com/vector-im/element-android/pull/5408/commits/ce2c309d72a357430ebc3f8685a3bb556622fd86